rosenpass/src/app_server.rs

```diff
@@ -549,7 +563,37 @@ impl AppServer {
                 }
 
                 ReceivedMessage(len, endpoint) => {
-                    match self.crypt.handle_msg(&rx[..len], &mut *tx) {
+                    let msg_result = match self.under_load {
+                        DoSOperation::UnderLoad { last_under_load: _ } => {
+                            //TODO: Lookup peer through addresses (hash)
+                            let index = self
+                                .peers
+                                .iter()
+                                .enumerate()
+                                .find(|(_num, p)| {
+                                    if let Some(ep) = p.endpoint() {
+                                        ep.addresses() == endpoint.addresses()
```

Comment:
Receiving a message from an unknown endpoint is actually all right…people would like to be able to connect to a VPN even when switching from mobile hotspot to office wifi :)

Also this triggers an expensive `O(n)` operation to find the address, which might actually be used for a DOS.

Lets just use the address in the IP packet…

The entire cookie mechanism is there to perform proof-of-ip-ownership, i.e. handle IP spoofing attacks.

Reply:
Are peers configured with an IP address or an IP address range in RP? The intent was to find a peer index (or PeerPtr)

Reply:
Peers are sometimes not configured with any IP address at all. Consider the case of a client/server style setup: The client may be connected through mobile tethering, often with a changing IP-Address while the server may be on a home network, using upnp and dyndns.

In this scenario, the client looks up the server's ip address and sends an InitHello, the server will have no IP address for the client configured; the first connection will happen based on the client's InitHello message – the server can not send an InitHello because they don't know the client IP.

Once the handshake completes successfully (and only if it completes successfully!) the server will cache the client's IP address. Some time later, the client may switch IP addresses, in which case the IP address cached by the server will be incorrect. The server will keep trying to reconnect with an incorrect IP address, but with an incorrect IP.

The client can re-initiate the connection from their new IP address by sending an InitHello message and completing the handshake.

In this case the handshake must be performed even though the stored IP is a mismatch.

Reply:
commented concern on main thread